README.md

```diff
@@ -1 +1,263 @@
+
+``` Rust
+Ipld::Assoc([
+    ("role", IPLD::String("user")),
+    (
+        "links",
+        Ipld::Array([
+            Ipld::Cid("bafkreidb2q3ktgtlm5yio7buj3sypyghjtfh5ernsteqmakf4p2c5bwmyi"),
+            Ipld::Cid("bafkreic75ydg5vkw324oqkcmqltfvc3kivyngqkibjoysdwiilakh4z5fe"),
+            Ipld::Cid("bafkreiffdiz6raf46zrr3b2usufgz5fo44aggmocz4zappr6khhhljcdpy"),
+        ]),
+    ),
+    (
+        "sig",
+        Ipld::Binary([
+            0xf2, 0xe7, 0xda, 0x4b, 0xdc, 0x37, 0x08, 0x63, 0x7c, 0x71, 0xb4, 0x13, 0x51, 0x2a,
+            0x0b, 0xd6, 0x2e, 0xde, 0x68, 0xa8, 0x96, 0x2d, 0x25, 0xec, 0x0f, 0x62, 0xdb, 0x65,
+            0x59, 0xaf, 0x33, 0xdc, 0xc5,
+        ]),
+    ),
+]);
+```
+
+Note that the IPLD parser has dropped the `role: "admin"` key.
+
+The `"sig"` field is then removed, and the remaining fields serialized to binary;
+
+``` Rust
+Ipld::DagJson::serialize(
+    Ipld::Assoc([
+        ("role", IPLD::String("user")),
+        (
+            "links",
+            Ipld::Array([
+                Ipld::Cid("bafkreidb2q3ktgtlm5yio7buj3sypyghjtfh5ernsteqmakf4p2c5bwmyi"),
+                Ipld::Cid("bafkreic75ydg5vkw324oqkcmqltfvc3kivyngqkibjoysdwiilakh4z5fe"),
+                Ipld::Cid("bafkreiffdiz6raf46zrr3b2usufgz5fo44aggmocz4zappr6khhhljcdpy"),
+            ]),
+        )
+    ])
+);
+```
+
+The signature is then checked against the above fields, which passes since there's only a `role: "user"` entry. The application then goes onto continue to use the original JSON with the `role: "admin"` entry.
+
+# 2 Safety
+
+Data that has already been parsed to an in-memory IPLD representation can be canonically encoded to a trivially: it has already been through a [parser / validator](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/).
+
+Data purporting to conform to an IPLD encoding (such as [DAG-JSON](https://ipld.io/specs/codecs/dag-json/spec/)) MUST be validated prior to signature verification. This can be as simple as round-trip decoding/encoding the JSON and checking that the hash matches. A validation error MUST be signalled if it does not match. 
+
+> [Implementers] may provide an opt-in for systems where round-trip determinism is a desireable [sic] feature and backward compatibility with old, non-strict data is unnecessary.
+>
+> — [DAG-JSON Spec](https://ipld.io/specs/codecs/dag-json/spec/)
+
+As it is critical for signatures guard against various attacks, the assumptions around canonical encoding MUST be enforced.
+
+## 2.1 Signing CIDs
+
+Rather than validating the inline IPLD, replacing the data with a CID link to the content MAY be used instead. Note while this is very safe (as it impractical to alter a signed hash), this approach mixes data layout with security, and may have a performance, disk, and networking impacts.
```

Comment:
Two points that may be worth calling out:
1) This allows you to cache validation by CID
2) BUT if you do so, you must take into account possible revocation of the signing keys used for the memoized signatures

I think (1) is obvious, and on its own probably not worth mentioning, but I worry that implementers may not make the leap to (2), and will risk accepting cached signatures from revoked keys.